File: LeetCode/src/dp/EditDistance.java
package dp;

public class EditDistance {
    public int minDistance(String word1, String word2) {
        int m = word1.length();
        int n = word2.length();

        // Create a DP table to store the minimum edit distance
        int[][] dp = new int[m + 1][n + 1];

        // Initialize the first row and first column of the DP table
        for (int i = 0; i <= m; i++) {
            dp[i][0] = i; // Minimum edit distance to convert word1[0...i-1] to an empty string
        }
        for (int j = 0; j <= n; j++) {
            dp[0][j] = j; // Minimum edit distance to convert an empty string to word2[0...j-1]
        }

        // Fill in the DP table
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (word1.charAt(i - 1) == word2.charAt(j - 1)) {
                    // Characters match, no operation needed
                    dp[i][j] = dp[i - 1][j - 1];
                } else {
                    // Choose the minimum of three operations: insert, delete, or substitute
                    dp[i][j] = 1 + Math.min(Math.min(dp[i - 1][j], dp[i][j - 1]), dp[i - 1][j - 1]);
                }
            }
        }

        // Return the minimum edit distance
        return dp[m][n];
    }

    public static void main(String[] args) {
        EditDistance solution = new EditDistance();

        // Test case
        String word1 = "horse";
        String word2 = "ros";
        System.out.println(solution.minDistance(word1, word2)); // Output: 3
    }
}

File: LeetCode_Python/dp/strings/EditDistance.py
class EditDistance:
    def minDistance(self, word1: str, word2: str) -> int:
        m, n = len(word1), len(word2)

        dp = [[0] * (n + 1) for _ in range(m + 1)]

        for i in range(m + 1):
            dp[i][0] = i
        for j in range(n + 1):
            dp[0][j] = j

        for i in range(1, m + 1):
            for j in range(1, n + 1):
                if word1[i - 1] == word2[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1]
                else:
                    dp[i][j] = min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]) + 1

        return dp[m][n]


# Time Complexity: O(m * n), where m is the length of word1 and n is the length of word2.
# Space Complexity: O(m * n), for the dp array used to store the edit distances.


if __name__ == "__main__":
    solution = EditDistance()

    word1 = "horse"
    word2 = "ros"
    print("Expected Output : 3")
    print("Actual Output : ", solution.minDistance(word1, word2))

    word1 = "intention"
    word2 = "execution"
    print("Expected Output : 5")
    print("Actual Output : ", solution.minDistance(word1, word2))
